### ISSUE TITLE
[RFE] podman-compose support

### ISSUE THREAD

**Issue by seant-me** [cockpit-project/cockpit-podman#247]
Having the ability to 'run' a container is great, however what if you want to
Use Cases:
1. You want to redeploy your container to latest image
2. You want to launch a container with a different version
3. You've forgotten how to launch a running container with working configuration (volumes, ports, etc.)
4. You want to launch multiple containers based on an image
5. You want to maintain complex config to run a container

There are many ways to accomplish this but docker-compose solves these problems. Podman-compose is the podman version of docker-compose. 

The docker-compose.yml files can be stored in ~/.config/... by default and user-defined allowable
It would be nice to have a textbox to write a new docker-compose file. Also, a validator. 
It would be nice to have a link to docker-compose getting started documentation (https://docs.docker.com/compose/gettingstarted/) or example (https://docs.docker.com/compose/compose-file/)

**Comment by marusak** [comment]
This was also requested for docker but we never did it.
It would be definitely doable, especially the first part. We have a nice path select component, so we can just allow user to find the compose file on the system and run it.

> It would be nice to have a textbox to write a new docker-compose file. Also, a validator.

This is a bit more out of scope. A proper file editor would be probably a full subproject. Just dropping content into `contenteditable` element and then reading it again does not seem like a proper solution.
But we should give this some thought.

**Comment by seant-me** [comment]
Validation is provided by docker-compose config[1]. It doesn't seem to exist in podman-compose as yet (or planned?) And I don't know if you would want to have a hard dependency on docker-compose. It's an option though.

[1] https://docs.docker.com/compose/reference/config/

**Comment by seant-me** [comment]
1. Just being able to have a list of docker compose files and being able to execute in the ui would be extremely helpful.
2. Being able to save a run container 'task' as a docker-compose file would be help. You could just add a checkbox to the run tab "save configuration as a docker-compose file" would be great too.

**Comment by DerInternets** [comment]
This would be a great feature. Currently i have to run portainer in parallel to do this - a unified Web UI would be awesome.

**Comment by czadikem** [comment]
Yeah is this going to be build in soon?

**Comment by garrett** [comment]
There is currently no timeframe for importing files (including compose files, among others) as definitions to get a container (or group of containers, including a pod) up and running, but it's something I'd like for us to add and it's something I've thought about and have written a little bit:

- https://github.com/cockpit-project/cockpit-podman/issues/735#issuecomment-1049028639
- https://github.com/cockpit-project/cockpit/discussions/16059#discussioncomment-1663195

Even exporting and importing "kube" definitions, with `podman generate kube POD_OR_CONTAINER_HERE` (which generates YAML) and `podman play kube YAML_FILE_HERE` would be useful for several of the reasons why someone would want podman-compose support. (Although not all. It would be great if we supported `podman-compose` for an "import" too.)

Right now, if you're happy with using a command line, you can use something like this kube approach. It's originally meant for using podman on one machine and kubernetes on another, to migrate between the two, but you can migrate containers and pods (groups) of containers between podman on both sides still. It'd still be great to have this kind of functionality in Cockpit-Podman, of course.

Additionally, I'd like for us to support a rebuild container option that would allow one to "edit" values to build the container again and replace it with the newer version. This would also cover several of the cases above, without having to export and import or mess with creating any files manually.

_**TL;DR**: I agree; it'd be great. We're not working on it right now, but it would be nice to solve these points in various ways, including `podman-compose` support._

**Comment by johndbritton** [comment]
@Garrett I'm interested in this feature and might have some cycles to put into it depending on difficulty. Any direction you can provide around what is likely to be accepted into the existing module would be helpful.

**Comment by loxoron218** [comment]
Any news on this? :)

I'd really love to use cockpit to import a long list of containers and finally replace portainer :D

**Comment by garrett** [comment]
We haven't been working on Compose support, though it would still be nice to have.

Instead, we've been working on something that addresses all the same points: systemd integration (also known as "quadlets").  This turns containers into systemd services, working for both system and standard user accounts (thanks to rootless containers).

If you have a Docker Compose or Podman Compose file (or other container creation formats like a kube file or even Podman/Docker command line), you can already use [`podlet`](https://github.com/containers/podlet) to convert between them. The `podlet` command works well today, and using it with Cockpit provides some systemd integration, turning containers into systemd services manageable in Cockpit's "Services" page.  Running containers also appear in Cockpit Podman.

The `podlet` command is packaged for many distributions and can even run from a container (system or rootless user container in Podman).

Some of us use this at home. I've used Compose files (even from external projects) and find the systemd-integrated container services a better solution, as they're part of the system (and treated like any other system service) instead of requiring manual commands. The "quadlet" systemd integration can even be configured to automatically update to the latest container (optionally).

We're improving systemd integration ("quadlets"), which will enhance Cockpit Podman in the future.  It already partially works today (with a little extra effort setting up the container service definition file ("quadlets"), which "podlet" helps with).

Even with systemd integration, Compose file support would be nice. However, systemd integration is the team's priority, and it addresses everything full Compose support would (except importing existing Compose files, which "podlet" already handles).

Note: The upstream naming is a little confusing ("quadlet" is the "code name" for systemd container integration, and [`podlet`](https://github.com/containers/podlet) is the conversion command).

For more on integrating containers with systemd, see: <https://www.redhat.com/en/blog/quadlet-podman> (and remember, you don't have to write the containerfile definition from scratch if you use `podlet`).

---

So, yes, we've been working on (and will continue to enhance) solving the features outlined in this issue (listed on the original post in the thread), but haven't had a chance to add in any form of "compose" support. It would be nice to have "compose" support still somehow (like importing compose files, for example), but our focus is better integration on the system, including the overlap of what "compose" does. If anyone wants to help us out with "compose" file importing, that would be nice. But if you're looking for a solution for an existing "compose" file today, then using "podlet" is likely ideal.

**Comment by Gloryandel** [comment]
I think the popularity of podman is somewhat hindered by Quadlet. When I first started using it, I was overwhelmed by the Quadlet configuration. Currently, I use podman-compose to generate pods, and use user-level systemd Unit to check dependencies and start pods in batches and selectively. The only drawback is that it requires `loginctl enable-linger <user>` to solve the problem of descendant lingering, including user-level systemd Unit.

**Comment by tomillr** [comment]
Any change of getting support for podman compose anytime soon?

I'm currently looking to move away from Unraid and OpenMediaVault, so podman compose would be a nice way for users like me to move on from Docker.
